frontend/catalyst/debug.py

```diff
@@ -16,63 +16,49 @@
 
 import builtins
 import os
+import sys
+from functools import partial
 
 import jax
 from jax.interpreters import mlir
 
 import catalyst
 from catalyst.compiled_functions import CompiledFunction
 from catalyst.compiler import Compiler
-from catalyst.jax_primitives import print_p
+from catalyst.pennylane_extensions import callback
 from catalyst.tracing.contexts import EvaluationContext
 from catalyst.tracing.type_signatures import filter_static_args, promote_arguments
 from catalyst.utils.filesystem import WorkspaceManager
 
 
-# pylint: disable=redefined-builtin
-def print(x, memref=False):
-    """A :func:`qjit` compatible print function for printing values at runtime.
-
-    Enables printing of numeric values at runtime. Can also print objects or strings as constants.
-
-    Args:
-        x (jax.Array, Any): A single jax array whose numeric values are printed at runtime, or any
-            object whose string representation will be treated as a constant and printed at runtime.
-        memref (Optional[bool]): When set to ``True``, additional information about how the array is
-            stored in memory is printed, via the so-called "memref" descriptor. This includes the
-            base memory address of the data buffer, as well as the rank of the array, the size of
-            each dimension, and the strides between elements.
+def print(fmt: str, *args, **kwargs) -> None:
+    """
+    This function has been modified from its original form in the JAX project at
+    https://github.com/google/jax/blob/5eeebf2c1829bf3c66c947b6e12464a779434e29/jax/_src/debugging.py#L269
+    version released under the Apache License, Version 2.0, with the following copyright notice:
```

Comment:
Let's please keep the original docstring :) But in addition, add documentation on how to use the format string syntax.

In addition @erick-xanadu, there is a question of should we keep the `memref` functionality. In which case, we can motivate keeping this implementation as `print_array_memref(x)`.

Reply:
I don't think there is a reason to keep the memref functionality unless team members are using it. @dime10 did you use it? Now you could implement printing out the address easily by doing the following, which in my opinion is not too bad to change when debugging and we can remove all of the other code.

```diff
diff --git a/frontend/catalyst/pennylane_extensions.py b/frontend/catalyst/pennylane_extensions.py
index 66248aac..fc757a6e 100644
--- a/frontend/catalyst/pennylane_extensions.py
+++ b/frontend/catalyst/pennylane_extensions.py
@@ -2474,6 +2474,7 @@ class CallbackClosure:
         jnpargs = []
         for void_ptr, ty in zip(flat_args, self.getLowLevelSignature):
             memref_ty = ctypes.cast(void_ptr, ty)
+            print(memref_ty)
             nparray = ranked_memref_to_numpy(memref_ty)
             jnparray = jnp.asarray(nparray)
             jnpargs.append(jnparray)
```

Reply:
Got it, thanks @erick-xanadu. I have no strong opinion here, happy to keep it if it's useful/the team is using it, but also happy to not.